File: src/train_bot/formatter.py
from __future__ import annotations

import datetime as dt
from typing import Sequence

from .transport_api import Departure


def format_departures(
    origin_name: str,
    destination_name: str | None,
    departures: Sequence[Departure],
    *,
    requested_time: dt.datetime | None = None,
) -> str:
    """Render a text summary of departures for Telegram."""

    if not departures:
        header = _format_header(origin_name, destination_name, requested_time)
        return f"{header}\nNo matching services found."

    lines = [_format_header(origin_name, destination_name, requested_time)]
    for idx, service in enumerate(departures, start=1):
        destination = destination_name or service.destination_name
        timing = _format_timing(service)
        platform = f"Platform {service.platform}" if service.platform else "Platform TBC"
        operator = service.operator_name or service.service
        calling_points = _format_calling_points(service)
        lines.append(
            "\n".join(
                [
                    f"{idx}. {origin_name} ➜ {destination}",
                    timing,
                    platform,
                    f"Operator: {operator}",
                    calling_points,
                ]
            )
        )

    return "\n\n".join(lines)


def _format_header(
    origin_name: str,
    destination_name: str | None,
    requested_time: dt.datetime | None,
) -> str:
    if destination_name:
        base = f"Next services from {origin_name} to {destination_name}"
    else:
        base = f"Next services departing {origin_name}"

    if not requested_time:
        return base

    return f"{base} after {requested_time.strftime('%H:%M on %d %b')}"


def _format_timing(service: Departure) -> str:
    expected = service.expected_departure_time
    aimed = service.aimed_departure_time
    status = service.status
    if expected and expected != aimed:
        return f"Due {aimed} (est. {expected}, {status})"
    return f"Due {aimed} ({status})"


def _format_calling_points(service: Departure) -> str:
    if not service.calling_points:
        return "Calling points data unavailable."

    points = [point.station_name for point in service.calling_points]
    return "Calling at: " + ", ".join(points[:6]) + ("…" if len(points) > 6 else "")

File: worker/src/formatter.ts
import type { Departure } from "./models";

export function formatDepartures(
  originName: string,
  destinationName: string | undefined,
  departures: Departure[],
  requestedTime?: Date
): string {
  if (!departures.length) {
    return `${formatHeader(originName, destinationName, requestedTime)}\nNo matching services found.`;
  }

  const lines: string[] = [formatHeader(originName, destinationName, requestedTime)];
  departures.forEach((service, index) => {
    const destination = destinationName ?? service.destinationName;
    const timing = formatTiming(service);
    const platform = service.platform ? `Platform ${service.platform}` : "Platform TBC";
    const operator = service.operatorName ?? "Unknown operator";
    const callingPoints = formatCallingPoints(service);

    lines.push(
      [
        `${index + 1}. ${originName} ➜ ${destination}`,
        timing,
        platform,
        `Operator: ${operator}`,
        callingPoints
      ].join("\n")
    );
  });

  return lines.join("\n\n");
}

function formatHeader(
  originName: string,
  destinationName: string | undefined,
  requestedTime?: Date
): string {
  const base = destinationName
    ? `Next services from ${originName} to ${destinationName}`
    : `Next services departing ${originName}`;

  if (!requestedTime) {
    return base;
  }

  const time = requestedTime.toLocaleTimeString("en-GB", {
    hour: "2-digit",
    minute: "2-digit",
    hour12: false
  });
  const date = requestedTime.toLocaleDateString("en-GB", {
    day: "2-digit",
    month: "short"
  });

  return `${base} after ${time} on ${date}`;
}

function formatTiming(service: Departure): string {
  const expected = service.expectedDepartureTime;
  const aimed = service.aimedDepartureTime;
  const status = service.status || "UNKNOWN";
  const aimedDisplay = aimed ?? "TBC";

  if (expected && aimed && expected !== aimed) {
    return `Due ${aimedDisplay} (est. ${expected}, ${status})`;
  }
  if (expected && !aimed) {
    return `Due ${expected} (${status})`;
  }
  return `Due ${aimedDisplay} (${status})`;
}

function formatCallingPoints(service: Departure): string {
  if (!service.callingPoints.length) {
    return "Calling points data unavailable.";
  }

  const names = service.callingPoints.map((point) => point.stationName);
  const summary = names.length > 6 ? `${names.slice(0, 6).join(", ")}…` : names.join(", ");
  return `Calling at: ${summary}`;
}
